### DISCUSSION TITLE
Export the finished binary and byproducts from the build folder

### DISCUSSION THREAD

**Discussion by nylnx** [zephyrproject-rtos/zephyr#56301] [Q&A] [answered]
I want to get the finished `zephyr.elf` and other post-build products like map file, disassembly etc. and copy them to a different, better location. Unfortunately there seems to be no elegant solution for this. It is possible to configure all these outputs in Kconfig, including the actual name for the `zephyr.elf`(!), but it all just gets dumped into the `build/zephyr` folder, together with a whole host of temp files and other things.

The only thing that can be somewhat elegantly exported is the elf binary itself with `$<TARGET_FILE:${logical_target_for_zephyr_elf}>` using the undocumented `${logical_target_for_zephyr_elf}`.

What have you done? Have you never thought about it, because you just didn't need it? How would you solve this?

Thanks in advance :)

**Comment by yashi** [comment, marked as answer]
Set `CONFIG_KERNEL_BIN_NAME` with a directory name.

```
$ grep CONFIG_KERNEL_BIN_NAME build/zephyr/.config
CONFIG_KERNEL_BIN_NAME="foo/zephyr"
$ ls build/zephyr/foo
zephyr.bin  zephyr.elf  zephyr.hex  zephyr.lst  zephyr.map  zephyr.stat
```

**Reply by nylnx** [reply]
I also looked at the code and haven't noticed that I could just do this. What a great find! This enables the use of the CMake `install(DIRECTORY ...)` command and solves this problem.

I will still file an issue about that... There should be an easy, documented way to get the output files you produce.

**Reply by nylnx** [reply]
Anyone also getting a linker error while trying this workaround:
You need to **manually** create the `foo` folder. It will not do that automatically, at least for me. Maybe @yashi had more luck with his particular setup.
This can be automated in the `ZephyrBuildConfig.cmake` file like this:

```
set(CONFIG_KERNEL_BIN_NAME "\"foo/zephyr\"" CACHE STRING "Zephyr Binary Output Name")
file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/zephyr/foo")
```

**Reply by yashi** [reply]
hmm.  That's odd.  Does the following command work for you?

```
$ ls
a.txt
$ cmake -E copy a.txt foo/a.txt
$ find
.
./a.txt
./foo
./foo/a.txt
$ cmake --version
cmake version 3.25.1
```

On my Linux development environment, `cmake -E copy` creates non-existing directory when specified.  [The doc](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-E-arg-copy) doesn't say about a single file copy without distination directory, though.

Also, I'd rather set `CONFIG_KERNEL_BIN_NAME` in `prj.conf`, instead.

```
echo 'CONFIG_KERNEL_BIN_NAME="foo/zephyr"' >> prj.conf
```

**Reply by nylnx** [reply]
The problem is not with CMake. It's `gcc`/`ld` that doesn't support outputting to a subdirectory that doesn't exist. So `zephyr.elf` just doesn't even get written and everything after that just fails. 

```
$ ls
program.c
$ gcc program.c -o bin/program
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: cannot open output file bin/program: No such file or directory
collect2: error: ld returned 1 exit status
$ mkdir bin
$ gcc program.c -o bin/program
$ ls bin
program
```
Your setup might be more forgiving, since this is a really arbitrary limitation...

And having the `CONFIG_KERNEL_BIN_NAME` in the CMake file is a deliberate choice. I don't want this to be a concern of the project itself and also my particular setup generates part of the build files, including the `ZephyrBuildConfig.cmake` and the installation code in the `CMakeLists.txt`, while `prj.conf` is a manually written file. So as long as I don't overwrite `CONFIG_KERNEL_BIN_NAME` in my project's Kconfig, my project doesn't even know it can be installed or has a different output path :)
